testsuite/tests/typing-gadts/name_existentials.ml

```diff
@@ -117,3 +117,108 @@ let f = function Pair (x, y : int * _) -> x + y
 type ('a, 'b) pair = Pair of 'a * 'b
 val f : (int, int) pair -> int = <fun>
 |}]
+
+
+(* #11891: allow naming more types *)
+(* We stillonly allow to name freshly introduced existentials *)
+
+type _ th =
+  | Thunk : 'a * ('a -> 'b) -> 'b th
+let f1 (type a) : a th -> a = function
+  | Thunk (type b) (x, f : b * (b -> _)) -> f x
+let f2 (type a) : a th -> a = function
+  | Thunk (type b c) (x, f : b * (b -> c)) -> f x
+[%%expect{|
+type _ th = Thunk : 'a * ('a -> 'b) -> 'b th
+val f1 : 'a th -> 'a = <fun>
+Line 6, characters 29-41:
+6 |   | Thunk (type b c) (x, f : b * (b -> c)) -> f x
+                                 ^^^^^^^^^^^^
+Error: This type annotation attempts to bind "c"
+       to the non-locally-abstract type "a".
+|}]
+(* Do not allow to deduce extra assumptions *)
+let ko1 (type a) : a th -> a = function
+  | Thunk (type b c) (x, f : b * (b -> c option)) -> f x
+[%%expect{|
+Line 2, characters 29-48:
+2 |   | Thunk (type b c) (x, f : b * (b -> c option)) -> f x
+                                 ^^^^^^^^^^^^^^^^^^^
+Error: This pattern matches values of type "b * (b -> c option)"
+       but a pattern was expected which matches values of type "b * (b -> a)"
+       Type "c option" is not compatible with type "a"
+|}]
+(* Can only name fresh existentials *)
+let ko2 = function
+  | Thunk (type b c) (x, f : b * (b -> c)) -> f x
+[%%expect{|
+Line 2, characters 29-41:
+2 |   | Thunk (type b c) (x, f : b * (b -> c)) -> f x
+                                 ^^^^^^^^^^^^
+Error: This pattern matches values of type "b * (b -> 'a)"
+       but a pattern was expected which matches values of type "b * (b -> 'b)"
+       This instance of "'a" is ambiguous:
+       it would escape the scope of its equation
```

Comment:
I removed the scoping constraint on the type variable, as it is not necessary (it will be enforced later anyway). This way, we get the same error as below. The wording might be better though.